src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,231 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <uv.h>
+
+#include "nvim/api/private/helpers.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/lua/executor.h"
+#include "nvim/vim.h"
+#include "nvim/version.h"
+
+#ifdef __linux__
+
+#define XLIB_NAME "libX11.so.6"
+typedef void *_XDisplay;
+typedef struct {
+    char data[18];
+} _XkbStateRec;
+typedef int _XStatus;
+
+enum { kLockedModsOffset = 9 };
+enum { _XkbUseCoreKbd = 0x0100 };
+#endif
+
+#ifdef WIN32
+#include <Winuser.h>
+#endif
+
+#ifdef __APPLE__
+#define Boolean Boolean_I_Dont_Care
+#include <IOKit/IOKitLib.h>
+#include <IOKit/IOReturn.h>
+#include <IOKit/hidsystem/IOHIDLib.h>
+#include <IOKit/hidsystem/IOHIDParameter.h>
+#include <CoreFoundation/CoreFoundation.h>
+#undef Boolean
+#endif
+
+enum {
+    kNumLock = 0x10,
+    kCapsLock = 0x02,
+};
+
+
+#if defined(__linux__)
+static bool xlib_opened = false;
+static uv_lib_t xlib;
+static _XDisplay *(*_XOpenDisplay)(void *);
+static int (*_XCloseDisplay)(_XDisplay *);
+static _XStatus (*_XkbGetState)(_XDisplay *, unsigned, _XkbStateRec *);
+
+/// A helper function to be used to load the libX11 and required functions
+static bool os_xlib_init(const char **const err)
+{
+  if (xlib_opened) {
+    return true;
+  }
+  const int udl_ret = uv_dlopen(XLIB_NAME, &xlib);
+  if (udl_ret != 0) {
+    *err = "Failed to open library libX11";
```

Comment:
Function should end here.

Reply:
Missing reason why library failed to open in error message.

Reply:
I thought using this err message will enhance the abstraction, suppose if the error message used was something so specific, then it may be hard to understand for the person who is using this function, but if the error is supposed for the core developers, I don't think it would be needed, because, the presence of error would be easily traced by `gdb` so, why is there a need, I thought that providing simple error message will be better